### PR TITLE
[REVIEW] Support groupby operations for decimal dtypes

### LINKED ISSUE #7687: [FEA] Support groupby aggregations w/ Decimal columns
**Example**:
```
import cudf
from cudf.core.dtypes import Decimal64Dtype

df = cudf.DataFrame({'val': [0.01, 0.02, 0.03], 'id': [0, 1, 2]})

df['dec_val'] = df['val'].astype(Decimal64Dtype(7,2))
df.groupby('id').sum()
```

**Result**: Note that the `dec_val` column is dropped entirely.
```

  | val
-- | --

0.03
0.01
0.02
```

### PR DESCRIPTION
**Pull request by vyasr**
This PR resolves #7687. It also does a bit of cleanup of the internals of the code base. There is more that I would like to do, but I'll probably punt everything to a future PR that I don't directly have to touch for this change in the interest of quickly resolving the issue.

I still need help determining why a few aggregations aren't working. The problems fall into two groups:
1. The `var` and `std` aggregations currently don't fail, but they always return columns filled with NULLs. I found the implementation of the dispatch for these methods in `variance.cu`/`compound.cuh`, and at least nominally it seems like these methods _are not_ currently supported because the corresponding `enable_if_t` is based on whether the type satisfies `std::is_arithmetic`, which decimal types will not. However, I'm not sure whether the problem is that this classification is incorrect and these types are actually supported by `libcudf`, or if there really isn't an implementation; I tried to find one, but there are a lot of different files related to aggregation and I'm sure I didn't find all of them. If we simply don't have an implementation, I can remove these from the list of valid aggregations.
2. The `mean`, `quantile`, and `median` aggregations all raise a `RuntimeError` from `binaryop.hpp`: "Input must have fixed_point data_type." I've traced the error to the Cython `GroupBy.aggregate` method, specifically the line where it calls through to the underlying `c_obj`'s `aggregate` method. The call stack in C++ is pretty deep after that, though, and I haven't yet been able to pinpoint whether the failure is a missing cast somewhere (i.e. `libcudf` thinks that the column is a floating point type when it's really not) or if the problem lies elsewhere.

**Update**
Thanks to @codereport, I've now marked all the above as unsupported operations. After some discussion with other devs I've also handled the other extended types. I still need to write tests, but I think this PR is ready for review in its current form to identify if I've missed anything in the implementation.

### CODE REVIEW

**Comment by codereport** [comment]
@vyasr None of the aggregations mentioned in 1) or 2) are currently supported for `cudf::group_by` with Decimal32/64 in libcudf.

This is the [master issue](https://github.com/rapidsai/cudf/issues/3556) for all of `fixed_point` work (I kept it up to date at least while I was spending most of my time on `fixed_point`). The two hash/sort `group_by` PRs are here: https://github.com/rapidsai/cudf/pull/7169/ and https://github.com/rapidsai/cudf/pull/7190/. Looks like `MIN`, `MAX`, `SUM` & `COUNT` are the only ops that are currently supported. More work will need to be done in order to support other ops. Especially the ops like `MEAN`, which are hard coded to only work with floating point accumulator times (aka the variables that store the temporary values).

**Comment by vyasr** [comment]
Thanks @codereport! That's helpful. I've updated the list of supported operations in this PR.

File: python/cudf/cudf/core/groupby/groupby.py [hunk 1]
```diff
@@ -590,6 +571,47 @@ def rolling(self, *args, **kwargs):
...
+        def func(x):
+            """Compute the {} of the group.""".format(func_name)
+            return getattr(x, func_name)(*args, **kwargs)
+
+        func.__name__ = func_name
+        return self.agg(func)
+
+    setattr(
+        GroupBy, key, functools.partialmethod(_agg_func_name_with_args, key)
+    )
```
Comment (shwina) [inline python/cudf/cudf/core/groupby/groupby.py:612]: Nice! I like this better than what we had before!

File: python/cudf/cudf/core/groupby/groupby.py [hunk 2]
```diff
@@ -685,15 +707,11 @@ def __init__(
...
-            if key in self.obj:
-                return self.obj[key].groupby(
-                    self.grouping, dropna=self._dropna, sort=self._sort
-                )
-            raise
+    def __getattr__(self, key):
+        if key in self.obj:
+            return self.obj[key].groupby(
+                self.grouping, dropna=self._dropna, sort=self._sort
+            )
```
Comment (shwina) [inline python/cudf/cudf/core/groupby/groupby.py]: Shouldn't we raise an `AttributeError` otherwise?

Comment (vyasr) [reply python/cudf/cudf/core/groupby/groupby.py]: I beat you to it :) See my last commit.

File: python/cudf/cudf/core/groupby/groupby.py [hunk 3]
```diff
@@ -590,6 +571,47 @@ def rolling(self, *args, **kwargs):
...
+    "median",
+    "nunique",
+    "collect",
+}
+
+
+# Dynamically bind the different aggregation methods.
+for key in _VALID_GROUPBY_AGGS:
+
+    def _agg_func_name_with_args(self, func_name, *args, **kwargs):
```
Comment (shwina) [inline python/cudf/cudf/core/groupby/groupby.py:596]: Can we not move this to outside the loop?

Comment (vyasr) [reply python/cudf/cudf/core/groupby/groupby.py:596]: Yes, we definitely could. The reason I hadn't done that is because I was instead trying to modify this function to not require the function name as input, but to instead just use the `key` directly in the `getattr` call inside. For some reason that didn't work, though: some of the `test_groupby.py` tests failed and the error wasn't something obvious, so I didn't take the time to dig further. Any ideas why that might be? My guess is something due to when the local variable is bound makes `getattr` pull the wrong attribute from `x` at runtime, but I'm not really sure.

Comment (vyasr) [reply python/cudf/cudf/core/groupby/groupby.py:596]: In any case, perhaps moving it outside the loop and calling it is the better option anyway.

Comment (shwina) [reply python/cudf/cudf/core/groupby/groupby.py]: > Any ideas why that might be? 

Looks like a case of [late binding](https://docs.python-guide.org/writing/gotchas/#late-binding-closures)?

Comment (shwina) [reply python/cudf/cudf/core/groupby/groupby.py]: Agree, I like what we have now

Comment (vyasr) [reply python/cudf/cudf/core/groupby/groupby.py]: Ah yes, thanks for finding a good link. I dealt with this once [a long time ago](https://bitbucket.org/glotzer/hoomd-blue/pull-requests/222) (can't believe it's been five years...) and forgot the exact rules. But good, the new version is a definite improvement. I'll work on tests next.

**Review by shwina — APPROVED** [review]
LGTM!

**Review by kkraus14 — APPROVED** [review]

**Comment by vyasr** [comment]
@codereport @shwina any idea what might cause the failure I'm currently seeing? The results don't appear to be completely junk memory: depending on the operation the output is either 0 or something unreasonably large, so it appears that the CentOS test is revealing something more systematic. Unless I'm doing something unstable in the test itself (very possible, right now it's a trivial aggregation and I should make it a little more realistic), it's hard to imagine that the problem is on the Python side of the code though.

**Comment by shwina** [comment]
Agreed - this looks like a bug in libcudf somewhere. It would be helpful if we can confirm that by writing a C++ gtest equivalent to the pytest.

**Comment by vyasr** [comment]
@kkraus14 @shwina just a heads-up, I have converted this PR back to a draft to avoid it getting merged before I can try to resolve the failures I observed on [CentOS here](https://gpuci.gpuopenanalytics.com/job/rapidsai/job/gpuci/job/cudf/job/prb/job/cudf-gpu-test/CUDA=10.2,GPU_LABEL=gpu,OS=centos7,PYTHON=3.8/1709/#showFailuresLink). In the event that the tests don't fail again after my most recent set of changes, I'd like to do my due diligence to avoid committing a Heisenbug to the code base. I meant to try to resolve this today, but ended up identifying and fixing a host of other bugs in the process. I'll return to this tomorrow.

**Comment by kkraus14** [comment]
> I can try to resolve the failures I observed on [CentOS here](https://gpuci.gpuopenanalytics.com/job/rapidsai/job/gpuci/job/cudf/job/prb/job/cudf-gpu-test/CUDA=10.2,GPU_LABEL=gpu,OS=centos7,PYTHON=3.8/1709/#showFailuresLink)

In the past when there was a CentOS only failure like this it was due to accidentally using the return value of a void function where for some reason in Ubuntu it was always 0 whereas in CentOS it was garbage memory.

File: python/cudf/cudf/core/column/column.py [hunk 4]
```diff
@@ -1840,10 +1840,12 @@ def as_column(
...
             if isinstance(pyarrow_array.type, pa.Decimal128Type):
-                dtype = cudf.Decimal64Dtype.from_arrow(pyarrow_array.type)
+                pyarrow_dtype = cudf.Decimal64Dtype.from_arrow(
+                    pyarrow_array.type
+                )
             else:
-                dtype = arbitrary.dtype
-            data = as_column(pyarrow_array, dtype=dtype)
+                pyarrow_dtype = arbitrary.dtype
+            data = as_column(pyarrow_array, dtype=pyarrow_dtype)
```
Comment (shwina) [inline python/cudf/cudf/core/column/column.py]: 👍 

While we're at it, let's use `pyarrow_type` for clarity. As PyArrow calls them _types_, not _dtypes_ :(

**Comment by vyasr** [comment]
After tracing this pretty deep down the stack @codereport helped me identify that these operations are actually not supported on cuda 10.x due to a [compiler bug](https://nvbugswb.nvidia.com/NvBugs5/SWBug.aspx?bugid=3186317&cp=). The C++ tests for this already exist and are explicitly disabled. In the interest of testing this to the extent possible, I've adding a conditional skip with pytest based on the `nvcc` version available.

**Comment by kkraus14** [comment]
> After tracing this pretty deep down the stack @codereport helped me identify that these operations are actually not supported on cuda 10.x due to a [compiler bug](https://nvbugswb.nvidia.com/NvBugs5/SWBug.aspx?bugid=3186317&cp=). The C++ tests for this already exist and are explicitly disabled. In the interest of testing this to the extent possible, I've adding a conditional skip with pytest based on the `nvcc` version available. 

Instead of checking the nvcc version we should be able to use the `rmm._cuda` module to get the CUDA version and skip on 10.1 / 10.2.

**Review by kkraus14 — APPROVED** [review]

File: python/cudf/cudf/tests/test_groupby.py [hunk 5]
```diff
@@ -331,28 +314,91 @@ def emulate(df):
...
     )
+
     check_dtype = False if func in _index_type_aggs else True
     assert_eq(got_df, expect_df, check_dtype=check_dtype)
 
 
+@pytest.mark.skipif(
+    rmm._cuda.gpu.runtimeGetVersion() < 11000,
+    reason="These aggregations are not supported on CUDA 10.x.",
+)
```
Comment (kkraus14) [inline python/cudf/cudf/tests/test_groupby.py:342]: Actually, instead of skipping these, can we have these throw if the CUDA version is unsupported? If someone is running on CUDA 10.x these APIs will return arbitrary garbage data silently, which is worse than throwing.

**Review by kkraus14 — CHANGES_REQUESTED** [review]

File: python/cudf/cudf/_lib/groupby.pyx [hunk 6]
```diff
@@ -240,6 +242,10 @@ def _drop_unsupported_aggs(Table values, aggs):
         elif (
                 is_decimal_dtype(values._data[col_name].dtype)
         ):
+            if rmm._cuda.gpu.runtimeGetVersion() < 11000:
+                raise RuntimeError(
+                    "Decimal aggregations are not supported on CUDA 10.x."
```
Comment (kkraus14) [inline python/cudf/cudf/_lib/groupby.pyx:247]: Can we reword this to say that it's only supported on CUDA >= 11.0 due to an nvcc compiler bug? Right now it feels arbitrary and like we're forcing users to upgrade without reason.

**Review by kkraus14 — APPROVED** [review]

**Comment by kkraus14** [comment]
@gpucibot merge
